core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/kopt/KOptListMove.java

```diff
@@ -0,0 +1,208 @@
+package org.optaplanner.core.impl.heuristic.selector.move.generic.list.kopt;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import org.optaplanner.core.api.domain.solution.PlanningSolution;
+import org.optaplanner.core.api.function.TriPredicate;
+import org.optaplanner.core.api.score.director.ScoreDirector;
+import org.optaplanner.core.impl.domain.variable.descriptor.ListVariableDescriptor;
+import org.optaplanner.core.impl.domain.variable.index.IndexVariableSupply;
+import org.optaplanner.core.impl.domain.variable.inverserelation.SingletonInverseVariableSupply;
+import org.optaplanner.core.impl.heuristic.move.AbstractMove;
+import org.optaplanner.core.impl.heuristic.move.Move;
+import org.optaplanner.core.impl.score.director.InnerScoreDirector;
+
+/**
+ *
+ * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
+ * @param <Node_> The value type
+ */
+public class KOptListMove<Solution_, Node_> extends AbstractMove<Solution_> {
+
+    private final ListVariableDescriptor<Solution_> listVariableDescriptor;
+    private final Object entity;
+    private final KOptDescriptor<Solution_, Node_> descriptor;
+    private final List<FlipSublistMove<Solution_>> equivalent2Opts;
+    private final int postShiftAmount;
+
+    public KOptListMove(ListVariableDescriptor<Solution_> listVariableDescriptor,
+            Object entity,
+            KOptDescriptor<Solution_, Node_> descriptor,
+            List<FlipSublistMove<Solution_>> equivalent2Opts,
+            int postShiftAmount) {
+        this.listVariableDescriptor = listVariableDescriptor;
+        this.entity = entity;
+        this.descriptor = descriptor;
+        this.equivalent2Opts = equivalent2Opts;
```

Comment:
I'm thinking... if this is empty, why even generate this move? Can we avoid it, without some major expenses elsewhere?

Reply:
It empty if the generated move creates multiple cycles and patching fails. When the selector detects multiple cycles, it attempts to patch the move to a higher k-opt by connecting two alternative cycles. This may fail. The selector does multiple attempts in it search. If it fails a certain number of times, it returns an infeasible move (since then the same return logic can be used, which is less error-prone then returning null and returning a `NoChangeMove` instead)